### ISSUE TITLE
Consider support for automatically supporting vcs tag generation

### ISSUE THREAD

**Issue by jpakkane** [jpakkane/meson#33]
A common requirement is to create a source file that has the release number or revision control tag as a string. So something like this:

const char *vcs_tag = "1.0.0-8e8bg2e";

This tag needs to be updated on every build.

Maybe we should have a command to help generating them semiautomatically. Something like

src = vcs_info(
input : 'versiontag.c.in',
output : 'versiontag.c',
fallback : '1.0')

Meson would then check in the background which vcs the system is using (git, hg, bzr, svn at least?), grab the revision and write the file. If the revision number has not changed, it needs to not touch the output file so Ninja's restat kicks in and does not rebuild if there are no other changes.

**Comment by axxel** [comment]
It seems you have changed your mind regarding the worth of introducing a new top-level command to support some kind of vcs based version generation support, which I welcome. (For a reference, in case someone should find this issue conversation here, there has been some discussion about this feature in #30 before.)

But I'd like to point out a few advantages my original proposal has over your solution:

 1. I had a two-phase approach: 1. vcs_info, which determined the information from the vcs repository and 2. a vcs_configure, which used the first to replace a string in a text file. Doing both in one go, like your approach, means you have to run the underlying vcs command twice if you'd like to configure two files (like some version.c file and maybe some NSIS or dpkg installer script file). Having a cleaner/smaller user interface, might well be better than saving another fraction of a second during build. So that seems to be a fair choice.
 1. In this form, none of the two meson users that expressed a need for this kind of feature (@stoggi and myself) could use it, because of your hard coded choice of git/hg parameters. @stoggi seemed to be interested in `git show-ref -s refs/heads/master` instead of `git describe`, while I need `--num` option of `hg identify`. I recommend to allow the user to supply an optional custom command for general flexibility here and also change the default of the hg version to use the `--num` option to get a revision number instead of a hash tag as version info. It would be more consistent with what the other default options (see git/svn) do.
 1. It allows the user to overwrite the default name of the string to be replaced (@VCS_TAG@ here).
 1. For the automatic vcs detection, it adds the proper dependencies to the internal vcs state files. This allows to use the feature without having to force the always_rebuild flag. (See the web-link in my comment on #30 for where I got that idea from). I tried to go for *optimal* speed here. You might not consider it worth it.
 1. It detects the presence of a vcs on configure time instead of build time, which makes more sense to me. It seems like 'the right place', it is (theoretically) faster, it allows to inform the user about it with some logging. Since this is a behind the scenes detection, I find the last point essential.
 1. It does not force the use of named parameters 'input' and 'output'. Other commands like `executable` don't do that either. I find that inconsistent and unnecessary boiler plate.

To sum it up: I propose to port those features from my solution to yours (except nr 1 and maybe 4). This would only amend your vcs_tag command slightly but would make it useful for stoggi and myself. The example usage would be:
````
version_c = vcs_tag('version.c', 'version.c.in', replace_string : '@VERSION@',
    command : ['git', 'describe', '--dirty'], fallback : '1.0.0')
````
This time, I'll not start hacking before you agreed to the changes ;).

**Comment by stoggi** [comment]
My use case was `git describe --dirty` but I recently discovered you can specify the dirty mask with `git describe --dirty=*` as my version string has a length xonstraing. I agree being able to specify your own VCS command would be neat.

Does this just turn it into a generic method of redirecting standard output of a command into a file at a specified tag, that is always rebuilt? Because that may have other uses. Maybe inserting the build date/time for example. Or the username of the machine that performed the build.

(Sorry I've been a bit inactive, I'm currently travelling around Seoul and Hong Kong, so its difficult to contribute at the moment).

**Comment by axxel** [comment]
@stoggi: yes, it would turn into that generic 'configure type' target. I called it `configure_target` in my last 'compromise' proposal in the #30 discussion.

I see two options going forward here:
 1. call it `configure_target`, making it really general but then it would not make sense to have some vcs related default behaviour in case there was not `command` parameter, meaning the command parameter would need to be mandatory and the vcs use case would always have to be explicit. It would also need another extension to support the svn case (as I've pointed out earlier) due to the multi-line output of the `svn info` command. I would have gone for an optional reg-ex, which should be general enough to select the part of the string that is interesting.
 1. stick with `vcs_tag` but allow the optional `command` and `replace_string` parameter, meaning basic vcs support out of the box but also the possibility of 'misusing' it for other purposes, like you suggested.

I could live with both. I think the first is cleaner with respect to the name and possible applications, while the second is more feature-rich and user friendly for the presumed 90% use case of the thing. I could live with both.

**Comment by jpakkane** [comment]
The case of a fully reconfigurable file generator is simple: that is what custom_target is for.

The point of vcs_info is to provide a simple, straightforward, don't-have-to-think-about-it solution for the 95% use case. Providing for a custom generator command is something we might need to do (thanks, Git, for providing a gazillion choices for everything).

That being said, could you explain the use case for providing your own replacement string? Are there cases where you can't just edit your source to use @VCS_TAG@?

As far as the input and output arguments go, they should stay so they mirror custom_command. I'm not perfectly happy with how executable etc work. If it were only up to me they would have only one positional argument and all sources would be in a sources keyword argument. However the current approach is common shorthand (CMake uses it for example) and so the decision was made to support that. However that should be the exception rather than the norm, which is the Python mantra of "explicit is better than implicit".

**Comment by axxel** [comment]
If I understood you correctly, you'd be willing to allow an optional custom command to be passed to vcs_tag? I'd really suggest to move the vcs detection from the build time to the configure time then because of what I already said, plus, if a custom command is provided, there is no auto-detection necessary.

My use for a customizable replacement string is the idea of being flexible in general in my particular case, I'd want to setup a meson.build file that can coexists with my old CMake stuff for the time being.

Regarding the mandatory naming of parameters: I guess you meant to refer to custom_target (instead of custom_command)? Please also see your wiki documentation of vcs_tag. I really like terse, clear and clutter free expressions, hence, I strongly prefer the executable way. Having pretty much every target definition starting with a 'output : ' string would be just boilerplate in my eyes. And now, that you have the shorthand for the main build targets, I'd prefer that option for every equivalent build rule, equivalent meaning 'one output and one or more inputs`. Consistency is a value in its own right, in my opinion. I don't have anything against the optional use of explicitly identifying the parameters, though. Also, given that you pointed out the reduced line count of your alternative meson build instructions for your Qt Creator build compared to their native qmake code, you seem to like meson to be compact (without getting cryptic, of course), as well.

I'd be happy to merge the vcs detection and custom command feature from our two versions into a new proposal.

**Comment by jpakkane** [comment]
Interestingly when Meson's language syntax was originally designed it was looser in the way you propose. The universal feedback I got from people was they preferred the current one (with explicit keyword arguments and such). This is one of those cases where a case can be made for either approach because it is a matter of preference rather than fact. To prevent this debate from going on forever (like the git-vs-hg and upstart-vs-systemd-vs-whatever battles), I'm going to bring out my project leader hat and decree that the current approach with explicit input and output keywords is the way to go for the moment. We can re-examine this once 1.0 comes around and we have more usage experience.

All that said I'm totally fine with adding a custom command and replacement string and will gladly review such proposals given the following requirements:

- test case 73 keeps working without any changes to it (it represents the 90% case, and is the simplest I could make it)
- if no custom command is defined, the current behavior is preserved
- the custom command can only be a command, not a shell pipeline
- when using a custom command, its output is taken as-is without any further processing (apart from chopping off \n)

The last one of these means that you can't use SVN for the custom command. This is totally fine, the default implementation is what everyone will use in practice and if it is not, then people can use custom_target.

And for @stoggi, this won't become a general replacement feature. Meson already provides that in the form of custom_target. As an example to generate a time stamp just write a script that creates it and use build_always : true.

**Comment by axxel** [comment]
So named input/output parameters it is. Quick question, though: does "if no custom command is defined, the current behavior is preserved" mean, the user observable behavior is preserved or the internal mechanism (like config-time vs. build-time vcs detection) is not changed in any way? (I am pretty sure you mean the former but just wanted to make really sure :)).

**Comment by jpakkane** [comment]
The observable behaviour. Though if the new implementation looks like a Rube Goldberg machine, getting it approved for merging might prove to be tricky.
